pkg/adapter/adapter.go

```diff
@@ -31,17 +36,141 @@ type envConfig struct {
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Read the Credentials.
+	rawURI, err := ioutil.ReadFile(a.credentialspath + "/URI")
+	if err != nil {
+		return fmt.Errorf("Unable to get MongoDb URI field: secretPath %s : %w", a.credentialspath+"/URI", err)
+	}
+	URI := string(rawURI)
+	if URI == "" {
+		return fmt.Errorf("MongoDb URI field is an empty string: URI %s : %w", URI, err)
+	}
+
+	// Create new Client.
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		return fmt.Errorf("Error creating mongo client: %w", err)
+	}
+	// Get dataSource
+	dataSource := getDataSource(client, a.database, a.collection)
+
+	// Connect to Client.
+	err = client.Connect(ctx)
+	if err != nil {
+		return fmt.Errorf("Error connecting to database: %s", err)
+	}
+	defer client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	stream, err := dataSource.Watch(ctx, mongo.Pipeline{})
+	if err != nil {
+		return fmt.Errorf("Error setting up changeStream: %s", err)
+	}
+	defer stream.Close(ctx)
+
+	// Watch and process changes.
+	a.processChanges(ctx, stream)
 	return nil
 }
+
+// Returns the appropriate dataSource: database or collection.
+func getDataSource(client *mongo.Client, db string, coll string) dataSource {
+	if coll != "" {
+		return client.Database(db).Collection(coll)
+	}
+	return client.Database(db)
+}
+
+// processChanges processes the new incoming change, creates a cloud event and sends it.
+func (a *mongoDbAdapter) processChanges(ctx context.Context, stream *mongo.ChangeStream) {
+	// For each new change recorded.
+	for stream.Next(ctx) {
+		var data bson.M
+		if err := stream.Decode(&data); err != nil {
+			a.logger.Desugar().Error("Error decoding the change stream", zap.Error(err))
+		}
+		// Create corresponding event.
+		event, err := makeCloudEvent(data)
+		if err != nil {
+			a.logger.Desugar().Error("Failed to create event", zap.Error(err))
+		}
+
+		// Send event using client.
+		ctx := cloudevents.ContextWithTarget(ctx, a.sinkurl)
+
+		// Send that Event.
+		if result := a.ce.Send(ctx, *event); cloudevents.IsUndelivered(result) {
+			a.logger.Desugar().Error("Failed to send event", zap.Any("result", result))
+		}
+	}
+	return
+}
+
+// makeCloudEvent makes a cloud event out of the change object recevied.
+func makeCloudEvent(data bson.M) (*cloudevents.Event, error) {
+	// Create Event.
+	event := cloudevents.NewEvent(cloudevents.VersionV1)
+
+	// Set cloud event specs and attributes. TODO: issue #43
+	// 		ID     -> id of mongo change object
+	// 		Source -> database/collection.
+	// 		Type   -> type of change either insert, delete or update.
+	//		Data   -> data payload containing either id only for
+	//                deletion or full object for other changes.
+	event.SetID(data["_id"].(bson.M)["_data"].(string))
+	event.SetSource(data["ns"].(bson.M)["db"].(string) + "/" + data["ns"].(bson.M)["coll"].(string))
```

Comment:
source is the EventSource you set as env variable... Can you print the string to see what it is? IIRC you mentioned it had a weird format. Just add it here in the review comment so that I can take a look

Reply:
Same with the one you are currently setting here, so that we can compare

Reply:
EventSource is `10.12.0.34/countries`
where countries is the db and the ip address is the IP address of the main replica set pod I think, I think we should mention the collection somewhere in the cloud event, right?

Reply:
yes, I think we should change this... 
we can do in a follow up